### PR TITLE
feat: Added affine cipher

### PR DESCRIPTION
**Pull request by dsmurrow**
#### Description of Change

An affine cipher is a linear transformation where encryption is of the form $E(x)=(ax+b) \text{ mod m }$ where $x$ is the character being encrypted and $m$ is the number of characters in the alphabet being used. Decryption is of the form $D(x) = a^{-1}(x-b) \text{ mod m}$ where $a^{-1}$ is the modular multiplicative inverse of $a$ modulo $m$, i.e., $a*a^{-1} \equiv 1 \text{ (mod m)}$.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:  added affine cipher

### CODE REVIEW

File: cipher/affine.c [hunk 1]
```diff
@@ -0,0 +1,198 @@
+/**
+ * @file
+ * @brief An [affine cipher](https://en.wikipedia.org/wiki/Affine_cipher) is a
+ * letter substitution cipher that uses a linear transformation to substitute
+ * letters in a message.
+ * @details Given an alphabet of length M with characters with numeric values
+ * 0-(M-1), an arbitrary character x can be transformed with the expression (ax
+ * + b) % M into our ciphertext character. The only caveat is that a must be
+ * relatively prime with M in order for this tranformation to be invertible,
```
Comment (CascadingCascade) [inline cipher/affine.c:9]:
```suggestion
 * relatively prime with M in order for this transformation to be invertible,
```

File: cipher/affine.c [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+ */
+void test_string(char *s, int a, int b)
+{
+    char *copy = malloc((strlen(s) + 1) * sizeof(char));
+    strcpy(copy, s);
+
+    affine_key_t key = {a, b};
+
+    affine_encrypt(copy, key);
+    assert(strcmp(copy, s) != 0);  // assert that the encryption worked
```
Comment (CascadingCascade) [inline cipher/affine.c:159]: I dissent, this assertion only proves the encryption altered the string, to prove that it actually worked it needs to be compared against an expected encrypted output

**Review by CascadingCascade — DISMISSED** [review]

File: cipher/affine.c (hunk 2 above)
Comment (dsmurrow) [reply cipher/affine.c:159]: Fair enough!

File: cipher/affine.c [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+int modular_multiplicative_inverse(unsigned int a, unsigned int m)
+{
+    int x[2] = {1, 0};
+    div_t div_result;
+
+    if (m == 0)
+        return 0;
+    a %= m;
+    if (a == 0)
+        return 0;
```
Comment (Panquesito7) [inline cipher/affine.c:56]: Best to use brackets here to make things cleaner.
```suggestion
    if (m == 0) {
        return 0;
    }
    a %= m;
    if (a == 0) {
        return 0;
    }
```

File: cipher/affine.c [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+    free(copy);
+}
+
+/**
+ * @brief Test multiple strings
+ *
+ * @returns void
+ */
+void tests()
```
Comment (Panquesito7) [inline cipher/affine.c:177]:
```suggestion
static void tests()
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: cipher/affine.c [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+ */
+#define Z95_CONVERSION_CONSTANT 32
+
+/**
+ * @brief a structure representing an affine cipher key
+ */
+typedef struct
+{
+    int a;  ///< what the character is being multiplied by
+    int b;  ///< what is being added after the multiplication with a
```
Comment (Panquesito7) [inline cipher/affine.c:36]: \`` should be used for variables and such to not get confused, IMO.
```suggestion
    int b;  ///< what is being added after the multiplication with `a`
```

File: cipher/affine.c [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+} affine_key_t;
+
+/**
+ * @brief finds the value x such that (a * x) % m = 1
+ *
+ * @param a number we are finding the inverse for
+ * @param m the modulus the inversion is based on
+ *
+ * @returns the modular multiplicative inverse of a mod m
```
Comment (Panquesito7) [inline cipher/affine.c:45]: I believe these two are variables/parameters.
```suggestion
 * @returns the modular multiplicative inverse of `a` mod `m`
```

File: cipher/affine.c [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    test_string("The quick brown fox jumps over the lazy dog",
+                "K7: .*6<4 =-0(1 90' 5*2/, 0):- +7: 3>%& ;08", 94, 0);
+    test_string(
+        "One-1, Two-2, Three-3, Four-4, Five-5, Six-6, Seven-7, Eight-8, "
+        "Nine-9, Ten-10",
+        "H&60>\\2*uY0q\\2*p4660E\\2XYn40x\\2XDB60L\\2VDI0 "
+        "\\2V6B6&0S\\2%D=p;0'\\2tD&60Z\\2*6&0>j",
+        51, 18);
+
+    printf("all tests have successfully passed!\n");
```
Comment (Panquesito7) [inline cipher/affine.c:195]:
```suggestion
    printf("All tests have successfully passed!\n");
```

**Review by Panquesito7 — COMMENTED** [review]
Almost there! 😄

File: cipher/affine.c [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    inverse.a += ALPHABET_SIZE;
+    inverse.a %= ALPHABET_SIZE;
+
+    inverse.b = -(key.b % ALPHABET_SIZE) + ALPHABET_SIZE;
+
+    return inverse;
+}
+
+/**
+ * @brief Encrypts character string s with key
```
Comment (Panquesito7) [inline cipher/affine.c:103]:
```suggestion
 * @brief Encrypts character string `s` with key
```

**Review by Panquesito7 — APPROVED** [review]
Awesome work. Thanks! 🚀

**Review by alexpantyukhin — APPROVED** [review]
Excellent!

**Pull request merged**